nb2/api.py

```diff
@@ -1,86 +1,150 @@
-from flask import Blueprint, jsonify, request
+from flask import Blueprint
+from flask_restful import abort, Api, fields, marshal, reqparse, Resource
+from sqlalchemy.exc import IntegrityError
 
-from nb2 import db
-from nb2.errors import conflict, does_not_exist_error, validation_error
 from nb2.models import Person, Quote
-from nb2.validators import Validators
 
-bp = Blueprint('api', __name__)
 
+bp = Blueprint("api", __name__)
+api = Api(bp)
 
-@bp.route('/')
-def hello():
-    return 'Hello'
 
+class PersonResourceBase(Resource):
 
-@bp.route('/people', methods=['GET'])
-def get_all_people():
-    return jsonify([person.serialize() for person in Person.query.all()])
+    def __init__(self, *args, **kwargs):
+        self.fields = {
+            "id": fields.Integer,
+            "slack_user_id": fields.String,
+            "first_name": fields.String,
+            "last_name": fields.String,
+            "quotes": fields.List(fields.String(attribute='content')),
+        }
 
+        self.parser = reqparse.RequestParser(bundle_errors=True)
 
-@bp.route('/people/<slack_user_id>', methods=['GET'])
-def get_person(slack_user_id):
-    person = Person.query.filter_by(slack_user_id=slack_user_id).first()
+        super(PersonResourceBase, self).__init__(*args, **kwargs)
 
-    if person is None:
-        error_msg = f"Person with slack_user_id {slack_user_id} does not exist"
-        return does_not_exist_error(error_msg)
 
-    return jsonify(person.serialize())
+class PersonResource(PersonResourceBase):
 
+    ERRORS = {
+        "does_not_exist": "Person with slack_user_id {slack_user_id} does not exist"
+    }
 
-@bp.route('/people', methods=['POST'])
-def create_person():
-    data = request.get_json() or {}
-    slack_user_id = data.get('slack_user_id')
+    def get(self, slack_user_id):
+        person = Person.query.filter_by(slack_user_id=slack_user_id).one_or_none()
 
-    required_field_errors = Validators.validate_required_fields_are_provided(Person, data)
-    if required_field_errors:
-        return required_field_errors
+        if person is None:
+            abort(404, message=self.ERRORS["does_not_exist"].format(slack_user_id=slack_user_id))
 
-    if Person.query.filter(Person.slack_user_id == slack_user_id).first():
-        error_msg = f"Person with slack_user_id {slack_user_id} already exists"
-        return conflict(error_msg)
+        return marshal(person, self.fields), 200
```

Comment:
Can you explain why we're using `marshal` here? From briefly reading the docs, it's supposed to return an ordereddict? So is this what's responsible for returning the keys in order?

Reply:
I think flask-restful actually will maintain the order of keys no matter what you pass it. So if we kept the `serialize` method on the `Person` class, we could have returned `person.serialize(), 200` and the dict keys would be ordered as they were in the serialize method. `marshal` is just a different way of serializing than we were doing before. 

Instead of us defining the `serialize` method, we just define a dictionary of field names to field types (`self.fields` here) and pass it into `marshal` with the data. One nice thing I like is that `marshal` will take as data either a single person, or a list of people and can handle serializing in either case. It also still allows a lot of flexibility for serialization (that we probably won't need) like nesting fields and computing fields that aren't actually on the model. All of that could still be done in our own `serialize` method, but this moves the responsibility of serializing to the API and not the model, which I kind of like.

Reply:
Thanks for the explanation! Sounds good then